x-pack/filebeat/input/httpjson/input.go

```diff
@@ -81,532 +48,145 @@ func newRetryLogger() *retryLogger {
-				// Pagination control using HTTP Body fields
-				ri, err = createRequestInfoFromBody(in.config.Pagination, common.MapStr(response), common.MapStr(mm), ri)
-				if err != nil {
-					return err
-				}
-				if ri == nil {
-					break
-				}
-				if err = in.applyRateLimit(ctx, header, in.config.RateLimit); err != nil {
-					return err
-				}
-				in.log.Info("Continuing with pagination to URL: ", ri.URL)
-				continue
-			}
-		}
-		break
-	}
-
-	if mm != nil && in.config.DateCursor.IsEnabled() {
-		in.advanceCursor(common.MapStr(mm))
+	_, err := net.DialTimeout("tcp", net.JoinHostPort(in.config.URL.Hostname(), port), time.Second)
+	if err != nil {
+		return fmt.Errorf("url %q is unreachable", in.config.URL)
 	}
 
 	return nil
 }
 
-func (in *HttpjsonInput) getURL() string {
-	if !in.config.DateCursor.IsEnabled() {
-		return in.config.URL
-	}
+// Run starts the input worker then returns. Only the first invocation
+// will ever start the worker.
+func (in *httpJSONInput) Run(ctx v2.Context, publisher stateless.Publisher) error {
+	log := ctx.Logger.With("url", in.config.URL)
 
-	var dateStr string
-	if in.nextCursorValue == "" {
-		t := timeNow().UTC().Add(-in.config.DateCursor.InitialInterval)
-		dateStr = t.Format(in.config.DateCursor.GetDateFormat())
-	} else {
-		dateStr = in.nextCursorValue
-	}
+	stdCtx := ctxtool.FromCanceller(ctx.Cancelation)
 
-	url, err := url.Parse(in.config.URL)
+	httpClient, err := in.newHTTPClient(stdCtx)
 	if err != nil {
-		return in.config.URL
+		return err
 	}
 
-	q := url.Query()
-
-	var value string
-	if in.config.DateCursor.ValueTemplate == nil {
-		value = dateStr
-	} else {
-		buf := new(bytes.Buffer)
-		if err := in.config.DateCursor.ValueTemplate.Execute(buf, dateStr); err != nil {
-			return in.config.URL
-		}
-		value = buf.String()
-	}
+	dateCursor := newDateCursorFromConfig(in.config, log)
 
-	q.Set(in.config.DateCursor.URLField, value)
+	rateLimiter := newRateLimiterFromConfig(in.config, log)
 
-	url.RawQuery = q.Encode()
+	pagination := newPaginationFromConfig(in.config)
 
-	return url.String()
-}
+	requester := newRequester(
+		in.config,
+		rateLimiter,
+		dateCursor,
+		pagination,
+		httpClient,
+		log,
+	)
 
-func (in *HttpjsonInput) advanceCursor(m common.MapStr) {
-	if in.config.DateCursor.Field == "" {
-		in.nextCursorValue = time.Now().UTC().Format(in.config.DateCursor.GetDateFormat())
-		return
+	if err := requester.processHTTPRequest(stdCtx, publisher); err != nil {
+		return err
 	}
 
-	v, err := m.GetValue(in.config.DateCursor.Field)
-	if err != nil {
-		in.log.Warnf("date_cursor field: %q", err)
-		return
-	}
-	switch t := v.(type) {
-	case string:
-		_, err := time.Parse(in.config.DateCursor.GetDateFormat(), t)
-		if err != nil {
-			in.log.Warn("date_cursor field does not have the expected layout")
-			return
-		}
-		in.nextCursorValue = t
-	default:
-		in.log.Warn("date_cursor field must be a string, cursor will not advance")
-		return
+	if in.config.Interval == 0 {
+		return nil
```

Comment:
Is Interval==0 a configuration error? Maybe this is exactly what some users want, (long-)poll the server.